source/tutorials/declarative-and-reproducible-developer-environments.rst

```diff
@@ -1,61 +1,151 @@
+.. _declarative-reproducible-envs:
+
 Declarative and reproducible developer environments
 ===================================================
 
-Nix can create reproducible environments given a declarative
-configuration called a Nix expression.
+In :ref:`ad-hoc-envs` tutorial we took a dive into providing shell
+environments for when we need a quick'n'dirty way of getting hold
+of some tools.
+
+In this tutorial we'll take a look how to create :term:`reproducible`
+shell environments given a declarative configuration file called a Nix expression.
+
+
+When are declarative shell environments useful?
+-----------------------------------------------
+
+This is the quickest approach to getting started with Nix:
+
+- single command to invoke it via ``nix-shell``
+- works across different operating systems (Linux / MacOS)
+- share the exact same environment with all developers
 
-Reproducible means you can share
-the configuration with others and guarantee that they are using the same
-software as you.
+Developer environments allow you to:
 
-To get started, make a new folder and create a file called ``shell.nix``
-with the following contents:
+- provide CLI tools, such as ``psql``, ``jq``, ``tmux``, etc
+- provide developer libraries, such as ``zlib``, ``openssl``, etc
+- set shell environment variables
+- execute bash during environment activation
+
+
+An example
+----------
+
+In top-level of your project create ``shell.nix`` with the following contents:
 
 .. code:: nix
 
-   { pkgs ? import <nixpkgs> {} }:
+   { pkgs ? import (fetchTarball "https://github.com/NixOS/nixpkgs/archive/3590f02e7d5760e52072c1a729ee2250b5560746.tar.gz") {} }:
 
    pkgs.mkShell {
      buildInputs = [
        pkgs.which
        pkgs.htop
+       pkgs.zlib
```

Comment:
What is the purpose of zlib in that example?

Reply:
To show off that not only CLI is useful but also libraries due to LD_LIBRARY_PATH. Maybe that should be clearer, but it's mentioned in the first section explaning the motivation.